src/bc_protocol/media_packet.rs

```diff
@@ -0,0 +1,381 @@
+    #[error(display = "Timeout")]
+    Timeout(#[error(source)] std::sync::mpsc::RecvTimeoutError),
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
+pub enum MediaDataKind {
+    VideoDataIframe,
+    VideoDataPframe,
+    AudioDataAac,
+    AudioDataAdpcm,
+    InfoData,
+    Invalid,
+    Continue,
+    Unknown,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct MediaData {
+    pub data: Vec<u8>,
+}
+
+impl MediaData {
+    pub fn body(&self) -> &[u8] {
+        let lower_limit = self.header_size();
+        let upper_limit = self.data_size() + lower_limit;
+        let len = self.len();
+        if !self.complete() {
+            unreachable!(); // An incomplete packet should be discarded not read, panic if we try
+        }
+        &self.data[lower_limit..upper_limit]
+    }
+
+    pub fn complete(&self) -> bool {
+        if self.len() < self.header_size() {
+            return false;
+        }
+        let full_data_len = self.len() - self.header_size();
+        full_data_len >= self.data_size() && full_data_len < self.data_size() + 8
+    }
+
+    pub fn expected_num_packet(&self) -> usize {
+        self.data_size() / CHUNK_SIZE + 1
+    }
+
+    pub fn header(&self) -> &[u8] {
+        let lower_limit = 0;
+        let upper_limit = self.header_size() + lower_limit;
+        &self.data[min(self.len(), lower_limit)..min(self.len(), upper_limit)]
+    }
+
+    pub fn header_size_from_kind(kind: MediaDataKind) -> usize {
+        match kind {
+            MediaDataKind::VideoDataIframe => 32,
+            MediaDataKind::VideoDataPframe => 24,
+            MediaDataKind::AudioDataAac => 8,
+            MediaDataKind::AudioDataAdpcm => 16,
+            MediaDataKind::InfoData => 32,
+            MediaDataKind::Unknown | MediaDataKind::Invalid | MediaDataKind::Continue => 0,
+        }
+    }
+
+    pub fn header_size_from_raw(data: &[u8]) -> usize {
+        let kind = MediaData::kind_from_raw(data);
+        MediaData::header_size_from_kind(kind)
+    }
+
+    pub fn header_size(&self) -> usize {
+        MediaData::header_size_from_raw(&self.data)
+    }
+
+    pub fn data_size_from_raw(data: &[u8]) -> usize {
+        let kind = MediaData::kind_from_raw(data);
+        match kind {
+            MediaDataKind::VideoDataIframe => MediaData::bytes_to_size(&data[8..12]),
+            MediaDataKind::VideoDataPframe => MediaData::bytes_to_size(&data[8..12]),
+            MediaDataKind::AudioDataAac => MediaData::bytes_to_size(&data[4..6]),
+            MediaDataKind::AudioDataAdpcm => MediaData::bytes_to_size(&data[4..6]),
+            MediaDataKind::InfoData => 0, // The bytes in MediaData::bytes_to_size(&data[4..8]) seem to be the size of the header
+            MediaDataKind::Unknown | MediaDataKind::Invalid | MediaDataKind::Continue => data.len(),
+        }
+    }
+
+    pub fn data_size(&self) -> usize {
+        MediaData::data_size_from_raw(&self.data)
+    }
+
+    pub fn pad_size_from_raw(data: &[u8]) -> usize {
```

Comment:
Most of these functions do not need to be `pub`, because they are not used in any module other than this one.

Reply:
Done